### PR TITLE
Introduce "setup dependencies"

### PR DESCRIPTION
**Pull request by edsko**
.. And finally, the pièce de résistance, introducing suppport for setup dependencies! This PR relies on #2500, #2504 and #2514. 

In this PR we add a new section to the `.cabal` file in which packages can specify explicit setup dependencies. When this section is absent (i.e., currently for all packages) we don't do anything different to what we do now, but when the section is present then that section is used to the dependencies of a setup script. 

For example, suppose we have two versions of package A available in our global (or user local, if not using sandboxes) DB. If we have a package whose setup script relies on version 0.1 of A, rather than version 0.2, it will fail to build because by default `cabal-install` just makes all packages visible when compiling a setup script, which means that it will be linked against the newer version.

However, we can now use an explicit setup dependencies stanza to address this:

```
custom-setup
  setup-depends:       base, Cabal, a ==0.1.0.0
```

Setup dependencies are treated as independent in the solver: we can pick different versions of packages for the setup script that we pick for the package proper, although we make an attempt not to do so (as explained in the [blog post about independent goals](http://www.well-typed.com/blog/2015/03/qualified-goals/)).

As a sanity check, I ran the solver after all these PRs against all (the latest versions of) all packages on Hackage (all 7889), and wherever we produce an install plan we produce the _same_ install plan, so that's reassuring. 

In addition, I also did some performance measurements, but there's hardly any difference in performance. The modified solver is _slightly_ slower, but barely so: here's a density plot of execution time in seconds; red is stock cabal-install 1.22, blue is the solver with all my PRs:

absolute

And here's a density plot of the relative performance of the two solvers; 1.0 meaning exactly equal, below 1.0 meaning the modified solver is faster, above 1.0 meaning the modified solver is slower:

relative

Basically, the conclusion -- performance-wise anyway -- is: no significant difference.

### CODE REVIEW

**Comment by dcoutts** [comment]
I think the status here is that @kosmikus is happy for this to be committed, and for his comments to be addressed after merging. Right @kosmikus?

The only other thing is a follow-up patch to address @23Skidoo's comment about cabal version checks.

Anything else? From my point of view this is ready to go in (and we'll follow up on the comments).

**Comment by dcoutts** [comment]
Ok @kosmikus confirmed to me that he's happy to merge now and that we can address his comments after.

@23Skidoo so feel free to merge, or give me an ack and I'll merge.

**Comment by 23Skidoo** [comment]
@dcoutts 
Sure, if you want to address comments in a later PR, then go ahead, just don't forget about them.

**Comment by dcoutts** [comment]
@23Skidoo promise :-) Thanks.

**Comment by 23Skidoo** [comment]
Yay!

**Comment by ezyang** [comment]
One thing that hasn't been discussed here, is what recommended backwards-compatible usage for people who don't have the newest version of Cabal? Example: https://github.com/ekmett/lens/issues/496#issuecomment-61388088 where people want to support older versions of GHC with the stock Cabal.
